### PR TITLE
add more metrics to `tes3vector(2|3|4)`

### PR DESCRIPTION
**Pull request by herbertgithubaccount**
this will add more metrics to the `tes3vector` classes. currently this includes:
* manhattan distance (L1): added to 2d,3d, and 4d vectors
* chebyshev distance (L-inf): added to 2d,3d, and 4d vectors
* regular distance (L2): added to 2d and 4d vectors
* xy-distance (i.e. distance ignoring z-coordinates): added to 3d vectors

documentation is still a WIP, but all the actual coding *should* be done.

### CODE REVIEW

File: MWSE/TES3Vectors.cpp [hunk 1]
```diff
@@ -68,6 +68,25 @@ namespace TES3 {
...
+		float dx = x - vec2->x;
+		float dy = y - vec2->y;
+		return sqrt(dx * dx + dy * dy);
+	}
+
+	
+	float Vector2::distanceChebyshev(const Vector2* vec2) const {
+		float dx = fabs(x - vec2->x);
+		float dy = fabs(y - vec2->y);
+		return std::max(dx,dy);
```
Comment (C3pa) [inline MWSE/TES3Vectors.cpp:81]: Can you insert space after a comma in argument lists to stay consistent with the rest of the code?

File: autocomplete/definitions/namedTypes/tes3vector2/distance.lua [hunk 2]
```diff
@@ -0,0 +1,8 @@
+return {
+	type = "method",
+	description = [[Calculates the distance to another vector in the standard way.]],
```
Comment (C3pa) [inline autocomplete/definitions/namedTypes/tes3vector2/distance.lua:3]: You can specify it's Euclidean distance. In addition, you can link to the related Wikipedia article. The following will be rendered as a link with text in [ ].
```md
[Euclidean distance](https://en.wikipedia.org/wiki/Euclidean_distance)
```

File: autocomplete/definitions/namedTypes/tes3vector2/distanceChebyshev.lua [hunk 3]
```diff
@@ -0,0 +1,17 @@
+return {
+	type = "method",
+	description = [[Calculates the distance to another vector, using the Chebyshev metric, which is defined as
+
+	math.max(math.abs(v1.x - v2.x), math.abs(v1.y - v2.y))
+
+This is useful for ensuring that the x and y coordinates between two vectors are all (independently) within a certain distance from each other.
+
+Here is a geometric description of the difference between the normal distance and the Chebyshev distance for two `tes3vector3`s  `v1` and `v2`:
```
Comment (C3pa) [inline autocomplete/definitions/namedTypes/tes3vector2/distanceChebyshev.lua:9]: two `tes3vector2`

File: autocomplete/definitions/namedTypes/tes3vector2/distanceChebyshev.lua [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+* If `v1:distance(v2) <= 1`, then `v2` is contained in a sphere around `v1` with radius 1 (i.e. diameter 2).
```
Comment (C3pa) [inline autocomplete/definitions/namedTypes/tes3vector2/distanceChebyshev.lua:10]: contained in a circle around

File: autocomplete/definitions/namedTypes/tes3vector2/distanceChebyshev.lua [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+* If `v1:distanceChebyshev(v2) <= 1`, then `v2` is contained within a cube centered around `v1`, where the cube has length 2.
```
Comment (C3pa) [inline autocomplete/definitions/namedTypes/tes3vector2/distanceChebyshev.lua:11]: within a square centered around `v1`, where the square sides have length 2.

File: autocomplete/definitions/namedTypes/tes3vector2/distance.lua [hunk 6]
```diff
@@ -0,0 +1,8 @@
...
+	arguments = {
+		{ name = "vec", type = "tes3vector3" },
```
Comment (C3pa) [inline autocomplete/definitions/namedTypes/tes3vector2/distance.lua:5]: tes3vector2

File: autocomplete/definitions/namedTypes/tes3vector2/distanceChebyshev.lua [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+]],
+	arguments = {
+		{ name = "vec", type = "tes3vector3" },
```
Comment (C3pa) [inline autocomplete/definitions/namedTypes/tes3vector2/distanceChebyshev.lua:14]: tes3vector2

File: autocomplete/definitions/namedTypes/tes3vector2/distanceChebyshev.lua [hunk 8]
```diff
@@ -0,0 +1,17 @@
+return {
+	type = "method",
+	description = [[Calculates the distance to another vector, using the Chebyshev metric, which is defined as
+
+	math.max(math.abs(v1.x - v2.x), math.abs(v1.y - v2.y))
+
+This is useful for ensuring that the x and y coordinates between two vectors are all (independently) within a certain distance from each other.
```
Comment (C3pa) [inline autocomplete/definitions/namedTypes/tes3vector2/distanceChebyshev.lua:7]: If you want, you can also link the Wiki article here. The comparison between the three distances here is nice:
https://en.wikipedia.org/wiki/Euclidean_distance#Generalizations
or link the article on Chebyshev's distance.

But this isn't necessary. I really appreciate your geometric description. You taught me something new today.

File: autocomplete/definitions/namedTypes/tes3vector4/distanceChebyshev.lua [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+	type = "method",
+	description = [[Calculates the distance to another vector, using the Chebyshev metric, which is defined as
+
+	math.max(math.abs(v1.x - v2.x), math.abs(v1.y - v2.y), math.abs(v1.z - v2.z), math.abs(v1.w - v2.w))
+
+This is useful for ensuring that the x, y, z, and w coordinates between two vectors are all (independently) within a certain distance from each other.
+
+Here is a geometric description of the difference between the normal distance and the Chebyshev distance for two `tes3vector3`s  `v1` and `v2`:
+* If `v1:distance(v2) <= 1`, then `v2` is contained in a sphere around `v1` with radius 1 (i.e. diameter 2).
+* If `v1:distanceChebyshev(v2) <= 1`, then `v2` is contained within a cube centered around `v1`, where the cube has length 2.
```
Comment (C3pa) [inline autocomplete/definitions/namedTypes/tes3vector4/distanceChebyshev.lua:11]: IMHO, it's okay to leave the example in 3D here.

File: autocomplete/definitions/namedTypes/tes3vector4/distanceManhattan.lua [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+In the two-dimensional case, the Manhattan metric can be thought of 
+as the distance that two taxis will have to travel if they're following a grid system.
+The formula for the Manhattan distance is
+
+	math.abs(v1.x - v2.x) + math.abs(v1.y - v2.y) + math.abs(v1.z - v2.z) + math.abs(v1.w - v2.w)
+
+This is useful for checking how far you'd actually have to move if you're only allowed to move along one axis at a time.
+]],
+	arguments = {
+		{ name = "vec", type = "tes3vector3" },
```
Comment (C3pa) [inline autocomplete/definitions/namedTypes/tes3vector4/distanceManhattan.lua:13]: tes3vector4

**Review by C3pa — CHANGES_REQUESTED** [review]
Have you tried to use the new methods 😄? You have added new methods to these classes in C++, but Lua doesn't know about these. You need to bind these to be available in Lua. See how other Lua bindings for these types are done in `TES3VectorsLua.cpp`.

**Review by C3pa — APPROVED** [review]

**Comment by herbertgithubaccount** [comment]
> Have you tried to use the new methods 😄? 

😅 this is embarrassing. i really dropped the ball on this one. i had a lot of confusion when getting the C++ development environment set up. i was eventually able to compile everything but couldn't figure out how to actually test things in morrowind once compiled. so i thought i would do my "testing" in a different project and then make sure everything compiled properly when ported into MWSE. (very foolish in hindsight).

but i figured out how to actually test those things today, and made sure to test everything before submitting the latest commit. sorry for the lousy testing procedure on my part. still very new to C++ development. i'll make sure to ask ahead of time if i have these sorts of problems in the future.

**Comment by C3pa** [comment]
It's all good, we are humans, and these things happen. 😃 Thanks Herbert.

**Comment by herbertgithubaccount** [comment]
> It's all good, we are humans, and these things happen. 😃 Thanks Herbert.

thanks for the understanding. i'm glad you were able to learn some geometry from all of this at least 😄 . i still need to triple check that i implemented all of your requested documentation changes. they were really helpful by the way. 

i like the idea of adding some visuals as well. is there a nice way to embed a picture into the descriptions, or would it be better to link to one?

**Comment by C3pa** [comment]
Markdown syntax to embed images is `image description`. We have one definition that has a picture: https://github.com/MWSE/MWSE/blob/master/autocomplete/definitions/global/tes3ui/getPalette.lua#L5. If the image license permits usage, you can add an image to the `docs/source/assets/images` folder. The other option is to embed an external image. That would require adding an attribution note below the image.

**Comment by herbertgithubaccount** [comment]
i double checked all of the documentation, and implemented your suggestion of adding links to the wikipedia pages. the pictures i was thinking of using are from wikipedia and belong to the wikimedia commons, so the licensing should be good. i thought about it a bit more and feel like it might be better to skip the images and keep the links to wikipedia, so that it's consistent with the other documentation and doesn't take up too much vertical scrolling space.

i can add the images directly if you'd prefer though. either way, it's nice to know about how to embed images for future reference

**Pull request merged**